unikernel.ml

```diff
@@ -155,40 +180,77 @@ struct
-                   [ Middleware.is_user_admin_middleware api_meth user ]
-                 else [])
-                @ (if email_verified && false (* TODO *) then
-                     [ Middleware.email_verified_middleware user ]
-                   else [])
-                @ Option.fold ~none:[]
-                    ~some:(fun csrf ->
-                      [ Middleware.csrf_verification user now csrf ])
-                    form_csrf
-                @ [ Middleware.auth_middleware user ]
-              in
-              Middleware.apply_middleware middlewares
+              Middleware.apply_middleware
+                (middlewares ~form_csrf user)
                 (fun reqd ->
-                  let cookie =
-                    { cookie with user_agent = Middleware.user_agent reqd }
-                  in
-                  let cookies =
-                    List.map
-                      (fun (cookie' : User_model.cookie) ->
-                        if String.equal cookie_value cookie'.value then cookie
-                        else cookie')
-                      user.cookies
-                  in
-                  let updated_user = User_model.update_user user ~cookies () in
-                  Store.update_user store updated_user >>= function
-                  | Ok () -> f user
+                  Store.update_cookie_usage store cookie user reqd >>= function
+                  | Ok () -> f ~json_dict user
                   | Error (`Msg err) ->
                       Logs.err (fun m -> m "Error with storage: %s" err);
-                      Middleware.http_response reqd ~title:"Error" ~data:err
+                      Middleware.http_response reqd ~title:"Error"
+                        ~data:(`String (String.escaped err))
                         `Not_found)
                 reqd)
 
+  let process_api_request ~json_dict ~token_value ~current_time f store reqd =
+    match Store.find_by_api_token store token_value with
+    | Some (user, token) ->
+        if User_model.is_valid_token token current_time then
+          Store.increment_token_usage store token user >>= function
+          | Ok () -> f ~json_dict user
+          | Error (`Msg err) ->
+              Middleware.http_response reqd ~title:"Error" ~data:(`String err)
+                `Not_found
+        else
+          Middleware.http_response reqd ~title:"Error"
+            ~data:
+              (`String
+                "Authorization token has expired. Please generate a new token \
+                 from your account dashboard.")
+            `Not_found
+    | None ->
+        Middleware.http_response reqd ~title:"Error"
+          ~data:
+            (`String
+              ("Invalid authorization token. User not found for token "
+             ^ token_value))
+          `Not_found
+
+  let authenticate ?(email_verified = true) ?(check_admin = false)
+      ?(api_meth = false) ?(check_csrf = false) ?(check_token = false) store
+      reqd f =
```

Comment:
I struggle with the logic below - esp. given the 5 optional boolean parameters.

Would it make sense to write it in english what should be done in this function? An attempt:

"Trying to access any endpoint, we need to check user permissions. Depending on the endpoint, the user may authenticate with a cookie or with a token (API request).

So, the main logic is:
  if token_is_sufficient then
    if check_for_token then Ok `Token else if check_cookie then Ok `Cookie else Error ()
  else
    if check_cookie then Ok `Cookie else Error ()

Now, depending on other parameters, more checks need to be done: CSRF if check_csrf is true and Ok `Cookie is used.

Also, check_admin must be respected for the endpoints requiring higher privileges.

The api_meth is used ???, email_verified is used ???"

But here, I see a first match on "api_authentication" and "check_token", where additional nesting is done (extract_json_body). In a different branch we get an error, in another branch we have conditionals, and further arguments are passed along....

It feels more like an `authentication_and_process_request_and_decode_json_and_check_for_privileges_and_maybe_use_token_authentication_and_error_with_either_text_plain_or_json` -- so I'd guess we need more declarative code: do one thing in a function, do not overload with lots of optional parameters (that tend to be forgotten), and compose these smaller functions.

Reply:
Exposure: consider we have a malicious person in the room who tries to get something with mollymawk. Now, to minimize the attack surface, we should avoid doing lots of user input handling before we verified that the request is auhenticated.

The reasoning: dealing with user input is hairy, and may contain bugs (leading to exceptions etc.). If we expose the input handling (e.g. json parsing) to a point when the user is auhenticated, 99.999% of attacks will vanish in the void since the request was not from an authenticated source.

Reply:
So at the moment we don't use `email_verified` is not used, since we don't yet send any verification emails. 

`check_admin` as you correctly mentioned is used to check if the user can access that endpoint.

We have methods that return a UI (such as opening the login page) and the rest of the methods return JSON (such as processing the login request). To differentiate between these two types of endpoints, we are using:
`api_meth` for those that return JSON and otherwise we assume this should return UI.

`check_csrf` is used to indicate which routes must contain a csrf token and `check_token` is used to indicate which routes are accessible via the API interface.

Without the `check_csrf` flag, most GET request will fail as these don't have any csrf in the request and without `check_token`, everything is accessible from the API.

I think these are all the labels. 

So what `authenticate` is currently doing is checking if there's an authorization header in the request and if `check_token` is true, this way we know this request needs to be processed as an API request and we then extract the request JSON body and process the request. If we do have an authorization header and `check_token` is false, it could mean the user is trying to access another route (maybe create a new token) and here we can say the route is inaccessible via the api,

Now if there's no Authorization header and check_token is false, we can assume this request is coming from the UI and here we first `extract_csrf_token` and then process the request.

Reply:
> Exposure: consider we have a malicious person in the room who tries to get something with mollymawk. Now, to minimize the attack surface, we should avoid doing lots of user input handling before we verified that the request is auhenticated.
> 
> The reasoning: dealing with user input is hairy, and may contain bugs (leading to exceptions etc.). If we expose the input handling (e.g. json parsing) to a point when the user is auhenticated, 99.999% of attacks will vanish in the void since the request was not from an authenticated source.

Thank you for this. I think you are absolutely correct. We should perform all the authentication steps before actually looking at what the request wants. I'll do some refactoring and ping you